File: src/parsing/Lexer.cpp
#include <algorithm>
#include <cassert>
#include <cstring>
#include <vector>
#include "Lexer.h"
#include "../text/Utf8.h"
#include "../io/printing/FmtPosition.h"

namespace mot
{
    Lexer::Lexer(const Console& console, FileContent& content) :
            _console(console),
            _content(content),
            _input(content.Data().get()),
            _size(content.Size()),
            _trivia(content, 0, 0)
    {
        _content.ResetLineMarkers();
        StartNewLine(0);
    }

    TokenType Lexer::PeekType()
    {
        return Peek().Type();
    }

    Token& Lexer::Peek()
    {
        if (_nextToken == nullptr)
            _nextToken = Lex();

        return *_nextToken.get();
    }

    UP<Token> Lexer::Advance()
    {
        if (_nextToken == nullptr)
            return Lex();

        return std::move(_nextToken);
    }

    UP<Token> Lexer::Consume(TokenType type)
    {
        if (Peek().Type() == type)
            return Advance();

        return nullptr;
    }

    static inline bool IsAlpha(char ch)
    {
        return ch >= 'a' && ch <= 'z' || ch >= 'A' && ch <= 'Z' || ch == '_';
    }

    static inline bool IsAlphaNumeric(char ch)
    {
        return IsAlpha(ch) || ch >= '0' && ch <= '9';
    }

    static inline bool IsNewLine(char ch)
    {
        return ch == '\r' || ch == '\n';
    }

    UP<Token> Lexer::Lex()
    {
        auto lastTokenType = _lastTokenType;

        if (lastTokenType == TokenType::EndOfInput)
        {
            _console.Error() << "Error: Parser scanned past the end of input\n";
            _console.Error() << FmtPosition(_content, _position).NoColumn().NoLine();
            return NewToken(TokenType::Error, 0);
        }

        ConsumeTrivia();

        auto pos = _position;

        if (pos >= _size)
            return LexEndOfInput();

        switch (lastTokenType)
        {
            case TokenType::StartOfInput:
            case TokenType::EndOfLine:
            case TokenType::Indent:
            case TokenType::Outdent:
            {
                if (_position != _lineStart + _lineSpaces)
                {
                    // the whitespace consumed at the start of the line doesn't match the number of spaces at the start of the line, which means there might be a
                    // tab character.
                    for (auto i = _lineStart; i < pos; i++)
                    {
                        if (_input[i] == '\t')
                        {
                            _console.Error() << "Error: Tab character used for indentation. Four spaces must be used for indentation.\n";
                            _console.Error() << FmtPosition(_content, _position);
                            return NewToken(TokenType::Error, 0);
                        }
                    }
                }

                // We only need to check for indentation changes immediately after the above tokens
                if (_indentLevel * Lexer::SPACES_PER_INDENT != _lineSpaces)
                {
                    return LexIndentation();
                }
                break;
            }
        }

        auto ch = _input[pos];

        switch (ch)
        {
            case ':':
                return NewToken(TokenType::Colon, 1);
            case '>':
                return NewToken(TokenType::GreaterThan, 1);
            case '"':
                return LexQuotedText();
            case '\r':
            case '\n':
                if (lastTokenType == TokenType::GreaterThan)
                    return LexBlockText();

                return LexEndOfLine(ch);
            default:
                if (lastTokenType == TokenType::Colon)
                    return LexLineText();

                if (IsAlpha(ch))
                    return LexWord();

                _console.Error() << "Error: Unexpected character\n";
                _console.Error() << FmtPosition(_content, _position);
                return NewToken(TokenType::Error, 1);
        }
    }

    void Lexer::ConsumeTrivia()
    {
        auto input = _input;
        auto pos = _position;
        auto size = _size;
        auto isComment = false;

        while (pos < size)
        {
            auto ch = input[pos];
            switch (ch)
            {
                case ' ':
                case '\t':
                    pos++;
                    continue;
                case '\r':
                case '\n':
                {
                    auto lastTokenType = _lastTokenType;
                    if (lastTokenType == TokenType::EndOfLine || lastTokenType == TokenType::StartOfInput || lastTokenType == TokenType::Outdent)
                    {
                        // The lexer isn't supposed to emit an EndOfLine token immediately after a StartOfInput, EndOfLine, or Outdent token.
                        pos += (ch == '\r' && pos + 1 < size && input[pos + 1] == '\n') ? 2 : 1;
                        pos += StartNewLine(pos);
                        isComment = false;
                        break;
                    }
                    goto DONE; // last token wasn't a StartOfInput, EndOfLine, or Outdent, so this new line is significant (don't consume it as trivia).
                }
                case '#':
                    isComment = true;
                    pos++;
                    continue;
                default:
                    if (isComment)
                    {
                        pos++;
                        continue;
                    }
                    goto DONE;
            }
        }

        DONE:
        _trivia = FileSpan(_content, _position, pos);
        _position = pos;
    }

    UP<Token> Lexer::LexEndOfInput()
    {
        assert(_position == _size);

        auto lastTokenType = _lastTokenType;
        if (lastTokenType == TokenType::GreaterThan)
            return LexBlockText();

        if (lastTokenType == TokenType::EndOfLine || lastTokenType == TokenType::Outdent)
        {
            if (_indentLevel > 0)
            {
                _indentLevel--;
                return NewToken(TokenType::Outdent, 0);
            }

            return NewToken(TokenType::EndOfInput, 0);
        }

        return NewToken(TokenType::EndOfLine, 0);
    }

    UP<Token> Lexer::LexIndentation()
    {
        assert(_lineSpaces != _indentLevel * Lexer::SPACES_PER_INDENT);

        auto newLevel = _lineSpaces / Lexer::SPACES_PER_INDENT;
        auto diff = (int32_t)newLevel - (int32_t)_indentLevel;

        if (diff < 0)
        {
            assert(_indentLevel > 0);
            _indentLevel--;
            return NewToken(TokenType::Outdent, 0);
        }

        if (diff > 0)
        {
            _indentLevel++;
            return NewToken(TokenType::Indent, 0);
        }

        _console.Error() << "Error: Misaligned indentation. Indents must be multiples of four spaces.\n";
        _console.Error() << FmtPosition(_content, _position);
        return NewToken(TokenType::Error, 0);
    }

    UP<Token> Lexer::LexEndOfLine(char currentChar)
    {
        assert(IsNewLine(currentChar));
        assert(currentChar == _input[_position]);

        auto length = 1u;
        if (currentChar == '\r')
        {
            auto pos = _position;
            if (pos + 1 < _size && _input[pos + 1] == '\n')
                length = 2;
        }

        auto token = NewToken(TokenType::EndOfLine, length);
        StartNewLine(_position);
        return token;
    }

    UP<Token> Lexer::LexWord()
    {
        assert(IsAlpha(_input[_position]));

        auto start = _position;
        auto pos = start + 1;
        auto size = _size;
        auto input = _input;

        while (pos < size)
        {
            auto ch = input[pos];

            if (!IsAlphaNumeric(ch))
                break;

            pos++;
        }

        auto length = pos - start;
        return NewToken(TokenType::Word, length, MotString(_content.Data(), start, length));
    }

    UP<Token> Lexer::LexLineText()
    {
        auto start = _position;
        auto pos = start + 1;
        auto size = _size;
        auto input = _input;
        auto lastNonWhitespace = start;

        while (pos < size)
        {
            auto ch = input[pos];

            if (IsNewLine(ch) || ch == '#')
                break;

            if (ch != ' ' && ch != '\t')
                lastNonWhitespace = pos;

            pos++;
        }

        auto length = lastNonWhitespace + 1 - start;
        return NewToken(TokenType::LineText, length, MotString(_content.Data(), start, length));
    }

    UP<Token> Lexer::LexQuotedText()
    {
        assert(_input[_position] == '"');

        auto start = _position;
        auto pos = start + 1;
        auto size = _size;
        auto input = _input;

        auto hasEscapes = false;
        auto resultLength = 0u;

        while (pos < size)
        {
            auto ch = input[pos];

            if (IsNewLine(ch))
                break;

            if (ch == '"')
            {
                auto value = GetQuotedLiteral(start, pos, resultLength, hasEscapes);
                return NewToken(TokenType::QuotedText, pos + 1 - start, value);
            }

            if (ch != '\\')
            {
                pos++;
                resultLength++;
                continue;
            }

            char32_t escapeResult;
            int consumed;
            if (ParseEscapeSequence(pos, escapeResult, consumed))
            {
                pos += consumed;
                resultLength += Utf8::EncodedSize(escapeResult);
                hasEscapes = true;
            }
            else
            {
                uint32_t lineNumber, lineStart;
                _content.PositionDetails(start, &lineNumber, &lineStart, nullptr);
                _console.Error() << "Error: Unsupported escape sequence in quoted text\n";
                _console.Error() << FmtPosition(_content, pos);
                return NewToken(TokenType::Error, pos - start);
            }
        }

        _console.Error() << "Error: Unterminated quoted-string\n";
        _console.Error() << FmtPosition(_content, pos - start);
        return NewToken(TokenType::Error, pos - start);
    }

    MotString Lexer::GetQuotedLiteral(uint32_t openQuote, uint32_t closeQuote, uint32_t resultLength, bool hasEscapes)
    {
        assert(closeQuote > openQuote);

        if (closeQuote - openQuote == 1)
            return MotString();

        if (!hasEscapes)
        {
            // no escapes sequences, so the resulting string is simply a literal substring on the original content
            return MotString(_content.Data(), openQuote + 1, closeQuote - openQuote - 1);
        }

        SP<char> result(new char[resultLength], std::default_delete<char[]>());
        auto resultPtr = result.get();
        auto input = _input;

        // the slow path... we need to loop through looking for escape sequences again
        auto ii = openQuote + 1; // start copying after the first double quote
        auto ri = 0u;
        while (ii < closeQuote)
        {
            assert(ri < (int)resultLength);

            if (input[ii] != '\\')
            {
                resultPtr[ri] = input[ii];
                ii++;
                ri++;
            }
            else
            {
                char32_t escapeResult;
                int consumed;
                ParseEscapeSequence(ii, escapeResult, consumed);
                ii += consumed;
                ri += Utf8::Encode(escapeResult, &resultPtr[ri]);
            }
        }

        assert (ii == closeQuote);
        assert (ri == resultLength);

        return MotString(result, 0, resultLength);
    }

    bool Lexer::ParseEscapeSequence(uint32_t position, char32_t& out_char, int& out_escapeLength)
    {
        auto size = _size;
        auto input = _input;

        assert(position < size);
        assert(input[position] == '\\');

        if (position + 1 >= size)
            return false;

        auto formatChar = (unsigned char)input[position + 1];
        switch (formatChar)
        {
            case '\'':
            case '"':
            case '?':
            case '\\':
                // simple case where the escape sequence just represents the literal value of the second character
                out_char = formatChar;
                out_escapeLength = 2;
                return true;
            case 'a':
                out_char = '\a';
                out_escapeLength = 2;
                return true;
            case 'b':
                out_char = '\b';
                out_escapeLength = 2;
                return true;
            case 'f':
                out_char = '\f';
                out_escapeLength = 2;
                return true;
            case 'n':
                out_char = '\n';
                out_escapeLength = 2;
                return true;
            case 'r':
                out_char = '\r';
                out_escapeLength = 2;
                return true;
            case 't':
                out_char = '\t';
                out_escapeLength = 2;
                return true;
            case 'v':
                out_char = '\v';
                out_escapeLength = 2;
                return true;
            case 'u':
            {
                auto digits = 0;
                char32_t value = 0u;
                for (auto i = position + 2; digits < 6 && i < size; i++, digits++)
                {
                    auto ch = input[i];
                    if (ch >= '0' && ch <= '9')
                    {
                        value = (value << 4u) | (ch - 0x30u);
                        continue;
                    }

                    ch &= ~0x20u; // to uppercase
                    if (ch >= 'A' && ch <= 'F')
                    {
                        value = (value << 4u) | (ch - 'A' + 10u);
                        continue;
                    }

                    break;
                }

                out_char = value;
                out_escapeLength = 2 + digits;
                return digits > 0;
            }
        }

        return false;
    }

    UP<Token> Lexer::LexBlockText()
    {
        assert(_position == _size || _input[_position] == '\r' || _input[_position] == '\n');

        auto start = _position;
        auto end = start;
        auto size = _size;
        auto input = _input;

        // figure out which (if any) following lines are part of the block text
        auto requiredSpaces = (_indentLevel + 1) * Lexer::SPACES_PER_INDENT;
        auto valueLength = 0u;

        struct CopyRange
        {
            uint32_t start, end;
        };
        std::vector<CopyRange> copyRanges;

        while (end < size)
        {
            // check if the next line is part of the block text
            // start by advancing past the new line
            auto lineStart = end + (input[end] == '\r' && end + 1 < size && input[end + 1] == '\n' ? 2 : 1);

            auto pos = lineStart;
            while (pos < size && input[pos] == ' ')
            {
                pos++;
            }

            auto padding = std::min(pos - lineStart, requiredSpaces);

            if (padding < requiredSpaces && pos < size && !IsNewLine(input[pos]))
            {
                // We didn't reach the required number of spaces, and we're not at the end of the line.
                // This means "end" already points to the end of the block text.
                break;
            }

            // Now we're sure the previous new-line was part of this block text, so let's consume it.
            StartNewLine(lineStart);

            // find the end of this line so we can start over
            while (pos < size && !IsNewLine(input[pos]))
            {
                pos++;
            }

            if (copyRanges.size() == 0)
            {
                valueLength = pos - (lineStart + padding);
                copyRanges.push_back(CopyRange{lineStart + padding, pos});
            }
            else
            {
                // make the last range include the new line sequence
                assert(copyRanges.back().end == end);
                valueLength += lineStart - end;
                copyRanges.back().end = lineStart;

                if (pos > lineStart)
                {
                    valueLength += pos - (lineStart + padding);
                    copyRanges.push_back(CopyRange{lineStart + padding, pos});
                }
            }

            end = pos;
        }

        MotString value;
        if (valueLength > 0)
        {
            auto vi = 0u;
            SP<char> valueStr(new char[valueLength], std::default_delete<char[]>());
            auto valueStrPtr = valueStr.get();

            for (auto it = copyRanges.begin(); it != copyRanges.end(); ++it)
            {
                auto rs = it->start;
                auto length = it->end - rs;
                if (length > 0)
                {
                    memcpy(&valueStrPtr[vi], &input[rs], length);
                }

                vi += length;
            }

            assert(vi == valueLength);
            value = MotString(valueStr, 0, valueLength);
        }

        return NewToken(TokenType::BlockText, end - start, value);
    }

    uint32_t Lexer::StartNewLine(uint32_t lineStart)
    {
        _content.MarkLine(lineStart);

        auto pos = lineStart;
        auto size = _size;
        auto input = _input;
        while (pos < size && input[pos] == ' ')
        {
            pos++;
        }

        _lineStart = lineStart;

        auto spaces = pos - lineStart;
        _lineSpaces = spaces;

        return spaces;
    }

    UP<Token> Lexer::NewToken(TokenType type, uint32_t length)
    {
        return NewToken(type, length, MotString());
    }

    UP<Token> Lexer::NewToken(TokenType type, uint32_t length, MotString value)
    {
        _lastTokenType = type;
        auto start = _position;
        auto end = start + length;
        _position = end;
        return UP<Token>(new Token(type, _trivia, FileSpan(_content, start, end), value));
    }
}

File: src/parsing/lexer.rs
use super::content;
use super::tokens::{Token, TokenType};
use unic_ucd_category::GeneralCategory;

pub type LexerResult<'a> = super::Result<Box<Token<'a>>>;

pub struct Lexer<'a> {
    pub file_content: &'a content::FileContent,
    ch_iter_: content::ContentIterator<'a>,
    leading_trivia_start_: usize,
    text_start_: usize,
    next_result_: Option<LexerResult<'a>>,
    is_initial_state_:  bool,
    is_end_state_: bool,
}

impl<'a> Lexer<'a> {
    pub fn new(file_content: &content::FileContent) -> Lexer {
        let mut lexer = Lexer {
            file_content,
            ch_iter_: file_content.iter(0),
            leading_trivia_start_: 0,
            text_start_: 0,
            next_result_: None,
            is_initial_state_: true,
            is_end_state_: false,
        };

        file_content.reset_line_markers();
        lexer.start_new_line(0);
        lexer.ch_iter_.next();

        lexer
    }

    pub fn peek_type(&mut self) -> TokenType {
        if let None = self.next_result_ {
            self.next_result_ = Some(self.lex());
        };

        if let Some(ref result) = self.next_result_ {
            return match *result {
                Ok(ref token) => token.token_type,
                Err(_) => TokenType::Error,
            }
        }

        panic!("MOT Bug: BufferedResult was still 'None' after calling lex().");
    }

    pub fn advance(&'a mut self) -> LexerResult {
        let next = self.next_result_.take();
        match next {
            Some(result) => result,
            None => self.lex(),
        }
    }

    fn start_new_line(&mut self, line_start: usize) {
        self.file_content.mark_line(line_start);
    }

    fn reset_iterator(&mut self, position: usize) {
        if position != self.position() {
            self.ch_iter_ = self.file_content.iter(position);
            self.next_char();
        }
    }

    fn next_char(&mut self) -> Option<char> {
        self.ch_iter_.next()
    }

    fn position(&self) -> usize {
        self.ch_iter_.position()
    }

    fn current_char(&self) -> char {
        self.ch_iter_.current_char()
    }

    fn char_at(&self, pos: usize) -> char {
        self.file_content.text()[pos..].chars().next().unwrap()
    }

    fn is_end_of_input(&self) -> bool {
        self.ch_iter_.is_end_of_input()
    }

    fn content_slice(&self, start: usize, end: usize) -> &str {
        &self.file_content.text()[start..end]
    }

    fn error(&self, pos: usize, message: String) -> super::error::ParsingError {
        let pos_details = self.file_content.position_details(pos);
        super::error::ParsingError::new(&pos_details, message)
    }

    fn err_result<T>(&self, pos: usize, message: String) -> super::Result<T> {
        Err(self.error(pos, message))
    }

    fn err_unexpected_char<T>(&self) -> super::Result<T> {
        self.err_result(self.position(), format!("Unexpected character: {}", self.current_char()))
    }

    fn create_start_of_input(&self) -> LexerResult {
        panic!("not impl")
    }

    fn new_token(&mut self, token_type: TokenType, value: Option<String>) -> LexerResult<'a> {
        // need to save trivia start and text start because they get overwritten by consume_trivia
        let leading_trivia_start = self.leading_trivia_start_;
        let text_start = self.text_start_;
        let text_end = self.position();
        self.consume_trivia();

        Ok(Box::new(Token::new(
            token_type,
            self.file_content,
            leading_trivia_start,
            self.position(),
            text_start,
            text_end,
            value,
        )))
    }

    fn consume_trivia(&mut self) -> super::Result<()> {
        let text_end = self.position();
        let mut split = text_end; // Point where it switches from trailing to leading trivia
        let mut is_original_line = true;
        let mut current_line_has_comment = false;

        while !self.is_end_of_input() {
            let ch = self.current_char();
            if ch == '#' {
                self.consume_comment();
                current_line_has_comment = true;
            } else if ch == '\n' || ch == '\r' {
                self.handle_new_line_sequence()?;
                if is_original_line || !current_line_has_comment {
                    split = self.position();
                    is_original_line = false;
                }
                current_line_has_comment = false;
            } else if is_whitespace(ch) {
                self.next_char();
            } else {
                break;
            }
        }

        if self.is_initial_state_ && is_original_line {
            // If this is the start of the file and the first line is not simply whitespace or a
            // comment, then treat that whitespace as leading trivia for the first token instead of
            // trailing trivia on the StartOfInput token.
            self.leading_trivia_start_ = text_end;
        } else {
            self.leading_trivia_start_ = split;
        }

        self.text_start_ = self.position();

        Ok(())
    }

    fn consume_comment(&mut self) {
        debug_assert_eq!(self.current_char(), '#');

        while let Some(ch) = self.next_char() {
            if is_new_line_char(ch) {
                break;
            }
        }
    }

    fn handle_new_line_sequence(&mut self) -> super::Result<()> {
        let ch = self.current_char();
        if ch == '\r' {
            // make sure the CR is immediately followed by LF
            let cr_pos = self.position();
            match self.next_char() {
                Some('\n') => Ok(()),
                _ => self.err_result(
                    cr_pos,
                    String::from("Carriage return (CR) was not followed by a line feed (LF)")),
            }?
        }

        debug_assert_eq!(self.current_char(), '\n', "Expected LF");
        self.next_char();
        self.file_content.mark_line(self.position());
        Ok(())
    }

    fn lex(&mut self) -> LexerResult<'a> {
        if self.is_initial_state_ {
            let result = self.new_token(TokenType::StartOfInput, None);
            self.is_initial_state_ = false;
            return result;
        }

        if self.is_end_of_input() {
            if self.is_end_state_ {
                // prevent accidental infinite loop
                panic!("MOT Bug: Parser attempted to read past end of input")
            }
            self.is_end_state_ = true;
            return self.new_token(TokenType::EndOfInput, None);
        }

        // trivia should already have been consumed at this point
        let ch = self.current_char();
        match ch {
            ':' => {
                self.next_char();
                self.new_token(TokenType::Colon, None)
            },
            '[' => {
                self.next_char();
                self.new_token(TokenType::OpenBracket, None)
            },
            ']' => {
                self.next_char();
                self.new_token(TokenType::CloseBracket, None)
            },
            '@' => {
                match self.next_char() {
                    Some('[') => {
                        self.next_char();
                        self.new_token(TokenType::AtOpenBracket, None)
                    },
                    Some('=') => self.lex_literal_value(),
                    None => self.err_result(self.position(), String::from("Unexpected end of file")),
                    _ => self.err_unexpected_char(),
                }
            },
            '=' => self.lex_value(),
            _ => {
                if is_identifier_char(ch) {
                    self.lex_identifier()
                } else {
                    self.err_unexpected_char()
                }
            },
        }
    }

    fn lex_literal_value(&mut self) -> LexerResult<'a> {
        debug_assert_eq!(self.char_at(self.text_start_), '@');
        debug_assert_eq!(self.current_char(), '=');

        let start = self.position() + 1;

        // figure out whether this is single-line or multi-line
        while let Some(ch) = self.next_char() {
            if !is_whitespace(ch) {
                break;
            }
        }

        if !self.is_end_of_input() {
            if is_new_line_char(self.current_char()) {
                // we ran into an end-of-line before any non-whitespace, so this is multi-line
                return self.lex_multiline_value();
            }

            // this is a single-line value, so find the end of the line
            while let Some(ch) = self.next_char() {
                if is_new_line_char(ch) {
                    break;
                }
            }
        }

        // we should now be at the end of a single-line value
        debug_assert!(self.is_end_of_input() || is_new_line_char(self.current_char()));

        let value = String::from(&self.file_content.text()[start..self.position()]);
        self.new_token(TokenType::Value, Some(value))
    }

    fn lex_multiline_value(&mut self) -> LexerResult<'a> {
        debug_assert!(is_new_line_char(self.current_char()));

        // consume current new line
        self.handle_new_line_sequence();

        // look for each subsequent line which starts with ==

        panic!("not impl")
    }

    fn lex_value(&mut self) -> LexerResult<'a> {
        debug_assert_eq!(self.current_char(), '=');

        // trim leading whitespace
        while let Some(ch) = self.next_char() {
            if !is_whitespace(ch) {
                break;
            }
        }

        let start = self.position();

        // find end of line
        if !self.is_end_of_input() && !is_new_line_char(self.current_char()) {
            while let Some(ch) = self.next_char() {
                if is_new_line_char(ch) {
                    break;
                }
            }
        }

        // trim trailing whitespace
        let untrimmed = &self.file_content.text()[start..self.position()];
        let mut end = untrimmed.len();
        for (i, ch) in untrimmed.char_indices().rev() {
            if !is_whitespace(ch) {
                break;
            }
            end = i;
        }

        let value = String::from(&untrimmed[0..end]);
        self.new_token(TokenType::Value, Some(value))
    }

    fn lex_identifier(&mut self) -> LexerResult<'a> {
        debug_assert!(is_identifier_char(self.current_char()));

        while let Some(ch) = self.next_char() {
            if !is_identifier_char(ch) {
                break;
            }
        }

        let ident = String::from(&self.file_content.text()[self.text_start_..self.position()]);
        self.new_token(TokenType::Identifier, Some(ident))
    }
}

fn is_identifier_char(ch: char) -> bool {
    if ch < 'a' {
        if ch < 'A' {
            return ch >= '0' && ch <= '9';
        }
        return ch <= 'Z' || ch == '_';
    }

    if ch <= 'z' {
        return true
    }

    if ch <= '\u{7F}' {
        return false; // we've covered all the ASCII cases already
    }

    match GeneralCategory::of(ch) {
        GeneralCategory::UppercaseLetter |
        GeneralCategory::LowercaseLetter |
        GeneralCategory::TitlecaseLetter |
        GeneralCategory::ModifierLetter |
        GeneralCategory::OtherLetter |
        GeneralCategory::LetterNumber |
        GeneralCategory::DecimalNumber |
        GeneralCategory::ConnectorPunctuation |
        GeneralCategory::NonspacingMark |
        GeneralCategory::SpacingMark |
        GeneralCategory::Format => true,
        _ => false,
    }
}

fn is_whitespace(ch: char) -> bool {
    let is_match = match ch {
        ' ' | // space
        '\u{0009}' | // HORIZONTAL TAB
        '\u{000b}' | // VERTICAL TAB
        '\u{000c}' | // FORM FEED
        '\u{0085}' | // NEXT LINE
        '\u{00a0}' // NO-BREAK SPACE
            => true,
        _ => false,
    };

    if is_match {
        return true;
    }

    match GeneralCategory::of(ch) {
        GeneralCategory::SpaceSeparator |
        GeneralCategory::LineSeparator |
        GeneralCategory::ParagraphSeparator => true,
        _ => false,
    }
}

fn is_new_line_char(ch: char) -> bool {
    ch == '\n' || ch == '\r'
}
